### PR TITLE
External 76

### PR DESCRIPTION
**Pull request by TJ-Everett**
Audit - Application Rules can unintentionally prevent token burning(21)

### CODE REVIEW

File: src/application/ProtocolApplicationHandler.sol [hunk 1]
```diff
@@ -94,28 +94,41 @@ contract ProtocolApplicationHandler is Ownable, AppAdministratorOnly, IApplicati
...
-                maxTxSizePerPeriodByRiskRuleId,
-                usdValueTransactedInRiskPeriod[_to],
-                _usdAmountTransferring,
-                lastTxDateRiskRule[_to],
-                riskScoreTo
-            );
-            // set the last timestamp of check
-            lastTxDateRiskRule[_to] = uint64(block.timestamp);
+            /// if rule is active check if the recipient is address(0) for burning tokens
+            if (_to != address(0)){
```
Comment (ShaneDuncan602) [inline src/application/ProtocolApplicationHandler.sol]: Shouldn't this "if" statement only be on the to address portion of this rule check? If someone is burning, the "from address" risk score should still be checked to make sure this wouldn't violate their risk imposed transaction limit.

File: src/application/ProtocolApplicationHandler.sol [hunk 2]
```diff
@@ -94,28 +94,41 @@ contract ProtocolApplicationHandler is Ownable, AppAdministratorOnly, IApplicati
...
+                usdValueTransactedInRiskPeriod[_to] = ruleProcessor.checkMaxTxSizePerPeriodByRisk(
+                    maxTxSizePerPeriodByRiskRuleId,
+                    usdValueTransactedInRiskPeriod[_to],
+                    _usdAmountTransferring,
+                    lastTxDateRiskRule[_to],
+                    riskScoreTo
+                );
+                // set the last timestamp of check
+                lastTxDateRiskRule[_to] = uint64(block.timestamp);
+            } else if (_to == address(0)) {
```
Comment (ShaneDuncan602) [inline src/application/ProtocolApplicationHandler.sol]: Oh, I see. You put it in  this else. This seems a little redundant. Unless I'm missing something, you should be able to just wrap the "to account" check in the if and that's all you need.

File: src/application/ProtocolApplicationHandler.sol [hunk 3]
```diff
@@ -128,11 +141,12 @@ contract ProtocolApplicationHandler is Ownable, AppAdministratorOnly, IApplicati
     function _checkAccessLevelRules(address _from, address _to, uint128 _usdBalanceValuation, uint128 _usdAmountTransferring) internal {
         uint8 score = appManager.getAccessLevel(_to);
         uint8 fromScore = appManager.getAccessLevel(_from);
-        /// Check if recipient is not AMM and then check sender access level
-        if (AccessLevel0RuleActive && !appManager.isRegisteredAMM(_from)) ruleProcessor.checkAccessLevel0Passes(fromScore);
-        /// Check if sender is not an AMM and then check the sender access level
-        if (AccessLevel0RuleActive && !appManager.isRegisteredAMM(_to)) ruleProcessor.checkAccessLevel0Passes(score);
-        if (accountBalanceByAccessLevelRuleActive) ruleProcessor.checkAccBalanceByAccessLevel(accountBalanceByAccessLevelRuleId, score, _usdBalanceValuation, _usdAmountTransferring);
+        /// Check if recipient is not AMM or address(0) and then check sender access level
+        if (AccessLevel0RuleActive && !appManager.isRegisteredAMM(_from) && _to != address(0)) ruleProcessor.checkAccessLevel0Passes(fromScore);
```
Comment (ShaneDuncan602) [inline src/application/ProtocolApplicationHandler.sol]: It should still check the from address regardless of the the "to address"

File: src/economic/ruleProcessor/ApplicationRiskProcessorFacet.sol [hunk 4]
```diff
@@ -42,25 +43,28 @@ contract ApplicationRiskProcessorFacet is IRuleProcessorErrors, IRiskErrors {
...
+                    /// maxBalance must be multiplied by 10 ** 18 to account for decimals in token pricing in USD
+                    if (total > uint(rule.maxBalance[i]) * (10 ** 18)) {
+                        revert BalanceExceedsRiskScoreLimit();
+                    } else {
+                        ///Jump out of loop once risk score is matched to array index
+                        break;
+                    }
+                }
+                unchecked {
+                    ++i;
```
Comment (ShaneDuncan602) [inline src/economic/ruleProcessor/ApplicationRiskProcessorFacet.sol:60]: perfect

File: src/economic/ruleProcessor/ERC20RuleProcessorFacet.sol [hunk 5]
```diff
@@ -42,18 +42,18 @@ contract ERC20RuleProcessorFacet is IRuleProcessorErrors, IERC20Errors {
...
         if (data.getTotalOracleRules() != 0) {
             try data.getOracleRule(_ruleId) returns (NonTaggedRules.OracleRule memory oracleRule) {
                 uint256 oType = oracleRule.oracleType;
                 address oracleAddress = oracleRule.oracleAddress;
-                /// White List type
+                /// Allow List type
                 if (oType == uint(ORACLE_TYPE.ALLOWED_LIST)) {
-                    if (!IOracle(oracleAddress).isAllowed(_address)) {
-                        revert AddressNotOnAllowedList();
+                    if (_address != address(0)) {
```
Comment (ShaneDuncan602) [inline src/economic/ruleProcessor/ERC20RuleProcessorFacet.sol:52]: Can you add a comment in here? Like, "If allow list oracle rule implemented, exempt address(0) to allow burning"

File: src/economic/ruleProcessor/ERC20RuleProcessorFacet.sol [hunk 6]
```diff
@@ -42,18 +42,18 @@ contract ERC20RuleProcessorFacet is IRuleProcessorErrors, IERC20Errors {
...
+                        if (!IOracle(oracleAddress).isAllowed(_address)) {
+                            revert AddressNotOnAllowedList();
+                        }
                     }
-
-                    /// Black List type
+                    /// Deny List type
```
Comment (ShaneDuncan602) [inline src/economic/ruleProcessor/ERC20RuleProcessorFacet.sol]: Can you add a comment in here? Like, "If deny list oracle rule implemented, any registered address is denied, including address(0)"

File: src/token/ProtocolERC20Handler.sol [hunk 7]
```diff
@@ -216,8 +216,13 @@ contract ProtocolERC20Handler is Ownable, ProtocolHandlerCommon, AppAdministrato
...
         if (transactionLimitByRiskRuleActive) {
-            ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _transferValuation);
-            ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreTo, _transferValuation);
+            if (_to != address(0)){
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _transferValuation);
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreTo, _transferValuation);
+            } else if (_to == address(0)) {
+                /// if recipient is address(0) this is a a burn and check the sender risk score only 
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _transferValuation);
+            }
```
Comment (ShaneDuncan602) [inline src/token/ProtocolERC20Handler.sol]: Kind of the same thing as above. This same logic can be accomplished with a single if around the "to" portion of the check. It will save some lines of code.

File: src/token/ProtocolERC721Handler.sol [hunk 8]
```diff
@@ -222,11 +222,18 @@ contract ProtocolERC721Handler is Ownable, ProtocolHandlerCommon, IAdminWithdraw
...
+            if (_to != address(0)){
+                /// if recipient is not address(0) check sender and recipient risk scores to ensure transaction limit is within rule limits
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _thisNFTValuation);
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreTo, _thisNFTValuation);
+            } else if (_to == address(0)) {
+                /// if recipient is address(0) this is a a burn and check the sender risk score only 
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _thisNFTValuation);
+            }
         }
+
```
Comment (ShaneDuncan602) [inline src/token/ProtocolERC721Handler.sol]: same

File: src/token/ProtocolHandlerCommon.sol [hunk 9]
```diff
@@ -81,25 +81,30 @@ abstract contract ProtocolHandlerCommon is IAppManagerUser, IOwnershipErrors, IZ
...
-            tokenAmount = (IToken(tokenList[i]).balanceOf(_account));
-            if (tokenAmount > 0) {
-                try IERC165(tokenList[i]).supportsInterface(0x80ac58cd) returns (bool isERC721) {
-                    if (isERC721 && tokenAmount >= _nftValuationLimit) totalValuation += _getNFTCollectionValue(tokenList[i], tokenAmount);
-                    else if (isERC721 && tokenAmount < _nftValuationLimit) totalValuation += _getNFTValuePerCollection(tokenList[i], _account, tokenAmount);
-                    else {
+        /// check if _account is zero address. If zero address we return a valuation of zero to allow for burning tokens when rules that need valuations are active. 
+        if (_account == address(0)){
+            return totalValuation; 
+        } else {
```
Comment (ShaneDuncan602) [inline src/token/ProtocolHandlerCommon.sol:87]: This is good thinking. This short circuit will save unnecessary gas, especially in an app with many tokens.

File: test/ApplicationERC20.t.sol [hunk 10]
```diff
@@ -568,6 +619,17 @@ contract ApplicationERC20Test is DiamondTestUtil, RuleProcessorDiamondTestUtil {
...
+
+        /// test that burn works when user has accessLevel above 0 
+        applicationCoin.burn(5 * (10**18)); 
+        /// test burn when rule active and user has access level 0 
+        vm.stopPrank();
+        vm.startPrank(accessTier);
+        appManager.addAccessLevel(rich_user, 0);
+
+        vm.stopPrank();
+        vm.startPrank(rich_user);
```
Comment (ShaneDuncan602) [inline test/ApplicationERC20.t.sol]: This would be invalid. If the from account is 0, they should not be able to burn when the accessLevel0 rule is active. I think I may have confused you with my answer to a question about this last week. Sorry about that. I went and looked at my answer and it was pretty vague.

File: test/ApplicationERC721.t.sol [hunk 11]
```diff
@@ -218,6 +218,9 @@ contract ApplicationERC721Test is DiamondTestUtil, RuleProcessorDiamondTestUtil
         // transfer to user1 to exceed limit
         vm.expectRevert(0x24691f6b);
         applicationNFT.transferFrom(user2, user1, 3);
+
+        /// test that burn works with rule 
+        applicationNFT.burn(3);
```
Comment (ShaneDuncan602) [inline test/ApplicationERC721.t.sol:223]: Probably need additional tests here to ensure that burning an amount of tokens that would put the from account below their allowed minimum would be prohibited.

File: test/ApplicationERC721.t.sol [hunk 12]
```diff
@@ -485,6 +510,16 @@ contract ApplicationERC721Test is DiamondTestUtil, RuleProcessorDiamondTestUtil
...
+
+        /// set price of token 5 below limit of user 2 
+        vm.stopPrank();
+        vm.startPrank(defaultAdmin);
+        nftPricer.setSingleNFTPrice(address(applicationNFT), 5, 14 * (10 ** 18));
+        /// test burning with this rule active 
+        /// transaction valuation must remain within risk limit for sender 
+        vm.stopPrank();
+        vm.startPrank(user2);
+        applicationNFT.burn(5);
```
Comment (ShaneDuncan602) [inline test/ApplicationERC721.t.sol]: We need additional testing to make sure the burn won't all the from address to get over their max transaction limit.

**Review by ShaneDuncan602 — CHANGES_REQUESTED** [review]
Just some fine tuning. Also, I think we need some tests in here to make sure that the zero address checks are working for general tags, risk, and access level.

File: src/application/ProtocolApplicationHandler.sol (hunk 2 above)
Comment (TJ-Everett) [reply src/application/ProtocolApplicationHandler.sol]: Great point. I'll condense these down.

File: test/ApplicationERC20.t.sol (hunk 10 above)
Comment (TJ-Everett) [reply test/ApplicationERC20.t.sol]: This helps alot. Thank you for clarifying this. 

No burns for accessLevel0 but if they have accessLevelWithdrawalLimits active for example that may allow for an amount at access0 and those burns would be allowed but count towards their limit?

Comment (ShaneDuncan602) [reply test/ApplicationERC20.t.sol]: Not sure that I understand that question. Essentially, if the rule currently checks for general tags, risk, or access levels on the "from address" we still want to perform those checks. We just want to skip the ones that can restrict based on address(0) not having that data associated to it. Lack of tags on address(0) should never restrict a burn.

File: src/token/ProtocolERC20Handler.sol [hunk 13]
```diff
@@ -216,8 +216,10 @@ contract ProtocolERC20Handler is Ownable, ProtocolHandlerCommon, AppAdministrato
         uint8 riskScoreTo = appManager.getRiskScore(_to);
         uint8 riskScoreFrom = appManager.getRiskScore(_from);
         if (transactionLimitByRiskRuleActive) {
-            ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _transferValuation);
-            ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreTo, _transferValuation);
+                ruleProcessor.checkTransactionLimitByRiskScore(transactionLimitByRiskRuleId, riskScoreFrom, _transferValuation);
```
Comment (oscarsernarosero) [inline src/token/ProtocolERC20Handler.sol]: incorrect indentation. No biggie tho.

**Review by oscarsernarosero — APPROVED** [review]
I think this is really good job Tayler.

File: src/token/ProtocolERC20Handler.sol (hunk 13 above)
Comment (TJ-Everett) [reply src/token/ProtocolERC20Handler.sol]: Fixed. Thank you!

**Review by ShaneDuncan602 — APPROVED** [review]
Looks great.

**Comment by ShaneDuncan602** [comment]
merged with external
